exporter/kafkaexporter/pdata_marshaler.go

```diff
@@ -125,44 +128,280 @@ type pdataTracesMarshaler struct {
 	marshaler            ptrace.Marshaler
 	encoding             string
 	partitionedByTraceID bool
+	maxMessageBytes      int
 }
 
-func (p *pdataTracesMarshaler) Marshal(td ptrace.Traces, topic string) ([]*sarama.ProducerMessage, error) {
-	var msgs []*sarama.ProducerMessage
-	if p.partitionedByTraceID {
-		for _, trace := range batchpersignal.SplitTraces(td) {
-			bts, err := p.marshaler.MarshalTraces(trace)
+func (p *pdataTracesMarshaler) processTrace(
+	trace ptrace.Traces,
+	topic string,
+	key []byte,
+	messages *[]*sarama.ProducerMessage,
+	messageChunks *[]*ProducerMessageChunks,
+	packetSize *int) error {
+
+	if trace.ResourceSpans().Len() == 0 {
+		return nil
+	}
+
+	bts, err := p.marshaler.MarshalTraces(trace)
+	if err != nil {
+		return fmt.Errorf("failed to marshal traces: %w", err)
+	}
+
+	msg := &sarama.ProducerMessage{
+		Topic: topic,
+		Value: sarama.ByteEncoder(bts),
+		Key:   sarama.ByteEncoder(key),
+	}
+
+	// Deriving version from sarama library
+	// https://github.com/IBM/sarama/blob/main/async_producer.go#L454
+	currentMsgSize := msg.ByteSize(2)
+
+	if (*packetSize + currentMsgSize) <= p.maxMessageBytes {
+		*packetSize += currentMsgSize
+		*messages = append(*messages, msg)
+		return nil
+	}
+
+	if len(*messages) > 0 {
+		*messageChunks = append(*messageChunks, &ProducerMessageChunks{*messages})
+		*messages = make([]*sarama.ProducerMessage, 0)
+		*packetSize = 0
```

Comment:
Please don't do this, passing pointers and trying to mutate them within sub methods creates a higher risk of causing race conditions or worse, panics.